src/Database/LSMTree/Internal/RunReaders.hs

```diff
@@ -104,35 +102,14 @@ data Reader m fhandle =
 
 type KOp m fhandle = (SerialisedKey, Entry SerialisedValue (BlobRef m fhandle))
 
--- | On equal keys, elements from runs earlier in the list are yielded first.
--- This means that the list of runs should be sorted from new to old.
 new :: forall h .
      HasFS IO h
   -> HasBlockIO IO h
-  -> Maybe WB.WriteBuffer
-  -> [Run IO (FS.Handle h)]
-  -> IO (Maybe (Readers IO (FS.Handle h)))
-new fs hbio = newAtOffsetMaybe fs hbio Nothing
-
--- | On equal keys, elements from runs earlier in the list are yielded first.
--- This means that the list of runs should be sorted from new to old.
-newAtOffset :: forall h .
-     HasFS IO h
-  -> HasBlockIO IO h
-  -> SerialisedKey  -- ^ offset
-  -> Maybe WB.WriteBuffer
-  -> [Run IO (FS.Handle h)]
-  -> IO (Maybe (Readers IO (FS.Handle h)))
-newAtOffset fs hbio offset = newAtOffsetMaybe fs hbio (Just offset)
-
-newAtOffsetMaybe :: forall h .
-     HasFS IO h
-  -> HasBlockIO IO h
   -> Maybe SerialisedKey  -- ^ offset
   -> Maybe WB.WriteBuffer
   -> [Run IO (FS.Handle h)]
   -> IO (Maybe (Readers IO (FS.Handle h)))
-newAtOffsetMaybe fs hbio mOffset wbs runs = do
+new fs hbio mOffset wbs runs = do
```

Comment:
I suggest that the internal API here uses a custom and strict maybe type for the offset, something like:
```
data OffsetKey = NoOffsetKey | OffsetKey !SerialisedKey
```
and then use `!mOffset` here.

That would propagate the strictness in the serialised key back through most of the layers of callers, or if not it'd only need a `!` pattern here an there to work. As is, using a `Maybe` we can't make the interior key strict so easily, nor propagate the strictness backwards to callers.

In the public API keep `Maybe`.

Reply:
Makes sense, and we don't even need Maybe in the public API, since we have two separate functions with/without offset.